### PR TITLE
Add an insertable streams demonstrator for encryption.

### PR DESCRIPTION
**Pull request by alvestrand**
**Description**
Demonstrate JS-layer encryption of a video call.

This consists of sending a video track to a receiver and a
"middlebox". The receiver has the key, the middlebox doesn't.

The receiver can see the video; the middlebox can only see it
if the key is blank.

The encryption itself is silly.

**Purpose**
Demonstrate how to set up an Insertable Streams usage.

### CODE REVIEW

File: src/content/peerconnection/endtoend-encryption/js/videopipe.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+  };
+  pc2.ontrack = function(e) {
+    if (receiveTransform) {
+      const transform = new TransformStream({
+        start() {},
+        flush() {},
+        transform: receiveTransform
+      });
+      const receiverStreams = pc2.getReceivers()[0].createEncodedVideoStreams();
```
Comment (murillo128) [inline src/content/peerconnection/endtoend-encryption/js/videopipe.js]: wouldn't it be better to just use the receiver from the event?
```
const receiverStreams= e.receiver.createEncodedVideoStreams();
```

File: src/content/peerconnection/endtoend-encryption/js/videopipe.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+  const sender = pc1.addTrack(stream.getVideoTracks()[0], stream);
+  if (sendTransform) {
+    const senderStreams = sender.createEncodedVideoStreams();
+    const senderTransformStream = new TransformStream({
+      start() {},
+      flush() {},
+      transform: sendTransform
+    });
+    senderStreams.readableStream
```
Comment (murillo128) [inline src/content/peerconnection/endtoend-encryption/js/videopipe.js:40]: would it be interesting to do the encrypt/decrypt on a worker thread by transfering the streams to it?

File: src/content/peerconnection/endtoend-encryption/js/main.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    const view = new DataView(chunk.data);
+    const checksum = view.getUint32(chunk.data.byteLength - 4);
+    if (checksum != 0xDEADBEEF) {
+      console.log('Corrupted frame received');
+      console.log(checksum.toString(16));
+    }
+    const newData = new ArrayBuffer(chunk.data.byteLength - 4);
+    const newView = new DataView(newData);
+    for (let i = 0; i < chunk.data.byteLength - 4; ++i) {
+      newView.setInt8(i, ~view.getInt8(i));
```
Comment (murillo128) [inline src/content/peerconnection/endtoend-encryption/js/main.js]: same here:
```
newView.setInt8(i, view.getInt8(i) ^ currentCryptoKey.charAt(i % currentCryptoKey.length));
```

File: src/content/peerconnection/endtoend-encryption/js/main.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+function encodeFunction(chunk, controller) {
+  if (currentCryptoKey) {
+    const view = new DataView(chunk.data);
+    // Any length that is needed can be used for the new buffer.
+    const newData = new ArrayBuffer(chunk.data.byteLength + 4);
+    const newView = new DataView(newData);
+
+    for (let i = 0; i < chunk.data.byteLength; ++i) {
+      newView.setInt8(i, ~view.getInt8(i));
```
Comment (murillo128) [inline src/content/peerconnection/endtoend-encryption/js/main.js]: how about xor'ing with the key?
```
newView.setInt8(i, view.getInt8(i) ^ currentCryptoKey.charAt(i % currentCryptoKey.length));
```

**Review by murillo128 — COMMENTED** [review]
Nice work! Just some comments for improvement.

File: src/content/peerconnection/endtoend-encryption/js/videopipe.js (hunk 2 above)
Comment (alvestrand) [reply src/content/peerconnection/endtoend-encryption/js/videopipe.js:40]: Yep. That's a later demo.
This doesn't work yet (the CL for making frames transferable is still in review), so I didn't include it at this time.

File: src/content/peerconnection/endtoend-encryption/js/main.js (hunk 4 above)
Comment (alvestrand) [reply src/content/peerconnection/endtoend-encryption/js/main.js]: Done!

File: src/content/peerconnection/endtoend-encryption/js/main.js (hunk 3 above)
Comment (alvestrand) [reply src/content/peerconnection/endtoend-encryption/js/main.js]: Done!

File: src/content/peerconnection/endtoend-encryption/js/videopipe.js (hunk 1 above)
Comment (alvestrand) [reply src/content/peerconnection/endtoend-encryption/js/videopipe.js]: Done!

File: src/content/peerconnection/endtoend-encryption/index.html [hunk 5]
```diff
@@ -0,0 +1,75 @@
+<!DOCTYPE html>
+<!--
+ *  Copyright (c) 2015 The WebRTC project authors. All Rights Reserved.
```
Comment (fippo) [inline src/content/peerconnection/endtoend-encryption/index.html:3]: nit: update the year

File: src/content/peerconnection/endtoend-encryption/index.html [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    </p>
+
+    <div id="buttons">
+        <button id="start">Start</button>
+        <button id="call" disabled>Call</button>
+        <button id="hangup" disabled>Hang Up</button>
+    </div>
+    <div id="status"></div>
+
+    <a href="https://github.com/webrtc/samples/tree/gh-pages/src/content/peerconnection/multiple-relay"
```
Comment (fippo) [inline src/content/peerconnection/endtoend-encryption/index.html:63]: nit: fix the url

**Comment by lgrahl** [comment]
I really like this. Can you point me to the spec? I have some suggestions.

Nits: `start` and `flush` of `TransformStream` are optional.

If I may add a suggestion: We know that even if there's a warning in caps that people will use this example in real world projects. So, how about we drop in [tweetnacl-js](https://github.com/dchest/tweetnacl-js) and do non-silly encryption? :slightly_smiling_face:

**Comment by murillo128** [comment]
version with iv + aes gcm using web crypto:

```
	let counter = 0;
	const roomKey = await getRoomKey(roomId,secret);
	async function encrypt(chunk, controller) {
		try {
			//Get iv
			const iv = new ArrayBuffer(4);
			//Create view, inc counter and set it
			new DataView(iv).setUint32(0,counter <65535 ? counter++ : counter=0);
			//Encrypt
			const ciphertext = await window.crypto.subtle.encrypt(
				{
					name: "AES-GCM",
					iv: iv
				},
				roomKey,
				chunk.data
			);
			//Crate new encoded data and allocate size for iv
			chunk.data = new ArrayBuffer(ciphertext.byteLength + 4);
			//Get uint view
			const data = new Uint8Array(chunk.data);
			//Copy iv
			data.set(new Uint8Array(iv),0);
			//Copy encruptedframe
			data.set(new Uint8Array(ciphertext),4);
			//Write
			controller.enqueue(chunk);
		} catch(e) {
		}
	}

	async function decrypt(chunk, controller) {
		try {
			//decrypt
			chunk.data =  await window.crypto.subtle.decrypt(
				{
				  name: "AES-GCM",
				  iv: new Uint8Array(chunk.data,0,4)
				},
				roomKey,
				new Uint8Array(chunk.data,4,chunk.data.byteLength - 4)
			);
			//Write
			controller.enqueue(chunk);
		} catch(e) {
		}
	}
```

**Comment by lgrahl** [comment]
Now I'm even more scared. :sweat_smile: I'm not well versed with AES-GCM, so, for the record, I have no idea if this is safe in the end but the IV part (which I think is the nonce in AES-GCM) doesn't look very random or like a safe nonce counter to me. Please, no unsafe examples.

**Comment by murillo128** [comment]
i think you should review aes-gcm then or how srtp works.

**Comment by lgrahl** [comment]
That is exactly **not** the answer I want to have in an example. Please, provide comments and references instead of "RTFM".

Please explain the ternary for the counter. IIRC, the nonce is 12 bytes, so the full 12 bytes should be used when counting. Adding a random part is probably an option, too, if the key may be reused.

**Comment by murillo128** [comment]
It's under work, just wait a bit ;)

**Comment by alvestrand** [comment]
lgrahl: https://github.com/alvestrand/webrtc-media-streams/blob/master/explainer.md is the most up to date spec (the spec in that repo is an older version).

**Comment by alvestrand** [comment]
I think I'll leave the "encryption" as totally bogus in this iteration. A subtly broken crypto is more dangerous than a totally bogus one - and I'm not competent to make sure something we cook up for the occasion isn't subtly broken.

**Comment by guidou** [comment]
lgtm

**Comment by lgrahl** [comment]
> A JS example is available in the webrtc/sample repository. A better example, closer to the SFrame implementation was provided by Sergio, but eventually not added, because an invited expert does not understand it, did not want to check against the specs, and could not wait for this blog post. oh well.

http://webrtcbydralex.com/index.php/2020/03/30/secure-frames-sframes-end-to-end-media-encryption-with-webrtc-now-in-chrome/

Hi Alex, I'll say it out loud: You're poison to the whole WebRTC community and one of the reasons why people stay the f*** away from it.

**Comment by agouaillard-cosmo** [comment]
Lenart. The intend was not to upset your anyone, and was thought to be factual. As a courtesy and sign of good faith, I have removed the part you judged offensive, like I would have if you had privately and directly contacted me with a request to do so. Now the only written trace left is your comment above.

However entitled you are to have an opinion, I would humbly request that you respect the code of conduct in the place where you choose to post. In this thread this is it: https://webrtc.org/support/code-of-conduct

I would also like to bring to your attention the severity of defamation laws in many countries, especially when it is not slander but libel. I would advice caution and control in the future.

**Comment by lgrahl** [comment]
Updated to concur with defamation laws (probably). I have to say, **you** of all persons reminding me of the CoC... wow.

**Pull request merged**